src/Npgsql/HardCancellationBlock.cs

```diff
@@ -0,0 +1,30 @@
+﻿using System;
+using System.Threading;
+
+namespace Npgsql
+{
+    readonly struct HardCancellationBlock : IDisposable
+    {
+        static readonly AsyncLocal<bool> _entered = new AsyncLocal<bool>();
+
+        public static bool Entered => _entered.Value;
+
+        readonly bool _outerBlock;
+
+        public HardCancellationBlock(bool outerBlock) =>_outerBlock = outerBlock;
+
+        public static IDisposable Enter()
```

Comment:
By returning an interface you cause boxing, so the struct won't be allocated in place. Return a struct like `List<T>.GetEnumerator` returns it or other methods.

```suggestion
        public static HardCancellationBlock Enter()
```